cmd/dp-dimension-importer/main.go

```diff
@@ -25,61 +42,123 @@ func (r responseBodyReader) Read(reader io.Reader) ([]byte, error) {
 	}
 
 	log.Debug("application configuration", log.Data{"config": cfg})
 
-	extractedEventConsumer, err := kafka.NewConsumerGroup(cfg.KafkaAddr, cfg.DimensionsExtractedTopic, log.Namespace, kafka.OffsetNewest)
-	if err != nil {
-		log.ErrorC("could not create consumer", err, nil)
-		os.Exit(1)
-	}
+	// Incoming kafka topic for instances to process
+	instanceConsumer := newConsumer(cfg.KafkaAddr, cfg.IncomingInstancesTopic, log.Namespace)
 
-	insertedEventProducer, err := kafka.NewProducer(cfg.KafkaAddr, cfg.DimensionsInsertedTopic, 0)
-	if err != nil {
-		log.ErrorC("kafka producer error", err, log.Data{"topic": cfg.DimensionsInsertedTopic})
-		os.Exit(1)
-	}
+	// Outgoing topic for instances that have completed processing
+	instanceCompleteProducer := newProducer(cfg.KafkaAddr, cfg.OutgoingInstancesTopic)
 
-	var neo4jClient *client.Neo4j
-	if neo4jClient, err = client.NewNeo4j(cfg.DatabaseURL, cfg.PoolSize); err != nil {
-		log.ErrorC("unexpected error while to create database connection pool", err, log.Data{
-			logKeys.URL:      cfg.DatabaseURL,
-			logKeys.PoolSize: cfg.PoolSize,
-		})
-		os.Exit(1)
-	}
+	// Outgoing topic for any errors while processing an instance
+	errorEventProducer := newProducer(cfg.KafkaAddr, cfg.EventReporterTopic)
+
+	neo4jClient := createNeo4jClient(cfg)
 
 	newDimensionInserterFunc := func() handler.DimensionRepository {
 		return repository.NewDimensionRepository(neo4jClient, map[string]string{})
 	}
 
-	datasetAPI := client.DatasetAPI{
-		ResponseBodyReader:  responseBodyReader{},
-		HTTPClient:          &http.Client{},
-		DatasetAPIHost:      cfg.DatasetAPIAddr,
-		DatasetAPIAuthToken: cfg.DatasetAPIAuthToken,
+	// ImportAPI HTTP client.
+	datasetAPICli := client.NewDatasetAPI(cfg.DatasetAPIAddr, cfg.DatasetAPIAuthToken, responseBodyReader{}, &http.Client{})
+
+	// Handler for dimensionsExtracted events.
+	eventHandler := handler.NewDimensionExtractedEventHandler(
+		newDimensionInserterFunc,
+		&repository.InstanceRepository{Neo4j: neo4jClient},
+		datasetAPICli)
+
+	// MessageProducer for dimensionsInsertedEvents
+	instanceCompletedProducer := message.NewInstanceCompletedProducer(
+		instanceCompleteProducer,
+		schema.InstanceCompletedSchema,
+	)
+
+	// Errors handler
+	errorEventHandler := &handler.ErrorHandler{
+		Producer:   errorEventProducer,
+		Marshaller: schema.ErrorEventSchema,
 	}
 
-	eventHandler := &handler.DimensionsExtractedEventHandler{
-		NewDimensionInserter: newDimensionInserterFunc,
-		InstanceRepository:   &repository.InstanceRepository{Neo4j: neo4jClient},
-		DatasetAPI:           datasetAPI,
+	// HTTP Health check endpoint.
+	healthcheck.NewHandler(cfg.BindAddr, cfg.ShutdownTimeout)
+
+	ctx, cancelConsumerLoop := context.WithCancel(context.Background())
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
+		log.Info(gracefulShutdownMsg, nil)
+		shutdownCTX, _ := context.WithTimeout(ctx, cfg.ShutdownTimeout)
+
+		cancelConsumerLoop()
```

Comment:
As I understand - you want to call cancelConsumerLoop()  just before os.Exit (after calling the close functions). cancelConsumerLoop() will clean up the resources used within the context for the timeout. If all the close functions complete before the timeout, you can use this function to dispose of the timer.

Reply:
Sorry I misread this the first time. What I was referring to is the function that is returned from context.WithTimeout. ATM it is being ignored, but it should be called at the end of the graceful shutdown.

Reply:
Now that I have interpreted it correctly - I see you are using cancelConsumerLoop() function to break out of the loop. The issue here I think is that nothing waits for the loop to be broken before continuing. If a message is currently being consumed then the loop may not be broken by the time it continues to close the underlying kafka consumer. Consider implementing a close method on the consumer that blocks until it has broken the loop. See kafka.consumer-group in go-ns for an example.

Reply:
Been looking into select statement and my understanding is that within a go routine only 1 select case can be executing at once.
https://tour.golang.org/concurrency/5
With that in mind there will never be a case where we are part way through consuming a message when the case for `ctx.Done` is triggered.